react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js

```diff
@@ -7,17 +7,52 @@ const initialState = {
 };
 
 function reducer(state = initialState, action) {
+  if (state.originalData.length === 0) {
+    state.originalData = state.books;
+  }
+  const arr = [];
   switch (action.type) {
-    case actions.GET_BOOKS: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_TO_CART: // TODO to implement the logic
-      return { ...state };
-    case actions.ADD_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.REMOVE_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.SEARCH_ITEM: // TODO to implement the logic
-      return { ...state };
+    case actions.GET_BOOKS:
+      return {
+        ...state,
+        books: action.payload
+      };
+    case actions.ADD_TO_CART:
+      return {
+        ...state,
+        bookSelected: [...state.bookSelected, action.payload]
+      };
+    case actions.ADD_ITEM:
+      state.bookSelected.forEach(book => {
+        if (book.id === action.payload) {
+          book.quantity += 1;
+        }
+      });
+      return {
+        ...state,
+        bookSelected: [...state.bookSelected]
+      };
+    case actions.REMOVE_ITEM:
+      state.bookSelected.filter(book => {
```

Comment:
check again the docs on filter (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter), see if you can find a better way to use it here. Note that it returns the filtered array and receives a filter function that should return a boolean. So its something like
```
filteredArray = array.filter(element => fulfillsFilterCondition ? true : false);